src/rust/src/x509.rs

```diff
@@ -55,7 +92,100 @@ fn parse_x509_extension(
+                    .call_method1("_init_without_validation", (data.0,))?
+                    .to_object(py),
+                GeneralName::DirectoryName(data) => {
+                    let py_list = pyo3::types::PyList::empty(py);
+                    for rdn in data {
+                        let py_set = pyo3::types::PySet::empty(py)?;
+                        for attribute in rdn {
+                            let oid = x509_module
+                                .call_method1("ObjectIdentifier", (attribute.type_id.to_string(),))?
+                                .to_object(py);
+                            let tag_enum = py
+                                .import("cryptography.x509.name")?
+                                .getattr("_ASN1_TYPE_TO_ENUM")?;
+                            let py_tag = tag_enum.get_item(attribute.value.tag().to_object(py))?;
+                            // TODO: is it safe to assume utf8? Current Python does.
+                            let py_data = std::str::from_utf8(attribute.value.data())
+                                .map_err(|_| asn1::ParseError::InvalidValue)?;
+                            let na = x509_module
+                                .call_method1("NameAttribute", (oid, py_data, py_tag))?
+                                .to_object(py);
+                            py_set.add(na)?;
+                        }
+                        let py_rdn = x509_module
+                            .call_method1("RelativeDistinguishedName", (py_set,))?
+                            .to_object(py);
+                        py_list.append(py_rdn)?;
+                    }
+                    let py_name = x509_module.call_method1("Name", (py_list,))?.to_object(py);
+                    x509_module
+                        .call_method1("DirectoryName", (py_name,))?
+                        .to_object(py)
+                }
+                GeneralName::UniformResourceIdentifier(data) => x509_module
+                    .getattr("UniformResourceIdentifier")?
+                    .call_method1("_init_without_validation", (data.0,))?
+                    .to_object(py),
+                GeneralName::IPAddress(data) => {
+                    let ip_module = py.import("ipaddress")?;
+                    let ip_addr = ip_module.call_method1("ip_address", (data,))?.to_object(py);
+                    x509_module
+                        .call_method1("IPAddress", (ip_addr,))?
+                        .to_object(py)
+                }
+                GeneralName::RegisteredID(data) => {
+                    let oid = x509_module
+                        .call_method1("ObjectIdentifier", (data.to_string(),))?
+                        .to_object(py);
+                    x509_module
+                        .call_method1("RegisteredID", (oid,))?
+                        .to_object(py)
+                }
+                // TODO: can we combine this with the EDIPartyName path?
```

Comment:
Right now I've got two separate paths for X400 and EDIPartyName because it's not clear to me if there's a way to get the implicit tag value with a `_` catch-all. However, this duplication won't have full coverage because we have no EDIPartyName vectors (and I don't want to create one!)

Reply:
Since these integers are kind of gibberish, I don't see a way to do this without paths for each. Or not providing the integer values.